test/e2e/user_workload_monitoring_test.go

```diff
@@ -965,227 +561,422 @@ func assertTenancyForRules(t *testing.T) {
+	}
+
+	if s.Annotations == nil {
+		s.Annotations = make(map[string]string)
+	}
+
+	s.Annotations["monitoring.openshift.io/grpc-tls-forced-rotate"] = "true"
+
+	if err := f.OperatorClient.CreateOrUpdateSecret(ctx, s); err != nil {
+		t.Fatalf("error saving grpc-tls secret: %v", err)
+	}
+
+	// We know the amount of expected secrets in forehand.
+	// We should not calculate it on-the-fly as the calculation could be racy.
+	//
+	// 1. openshift-monitoring/prometheus-k8s-grpc-tls-[hash]
+	// 2. openshift-user-workload-monitoring/prometheus-user-workload-grpc-tls-[hash]
+	// 3. openshift-monitoring/thanos-querier-grpc-tls-[hash]
+	// 4. openshift-user-workload-monitoring/thanos-ruler-grpc-tls-[hash]
+	//
+	// The central grpc-tls secret is verified independently by getting it directly
+	// and verifying if the force-rotation annotation has been removed.
+	const expectedGRPCSecretCount = 4
+
+	err = framework.Poll(time.Second, 5*time.Minute, func() error {
+		s, err := f.KubeClient.CoreV1().Secrets(f.Ns).Get(ctx, "grpc-tls", metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("error loading grpc-tls secret: %v", err)
+		}
+
+		if _, ok := s.Annotations["monitoring.openshift.io/grpc-tls-forced-rotate"]; ok {
+			return errors.New("rotation did not execute: grpc-tls-forced-rotate annotation set")
+		}
+
+		got := countGRPCSecrets(f.Ns) + countGRPCSecrets(f.UserWorkloadMonitoringNs)
+		if expectedGRPCSecretCount != got {
+			return errors.Errorf("expecting %d gRPC secrets, got %d", expectedGRPCSecretCount, got)
+		}
+
+		return nil
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+// setupUWM creates the required dependencies to facilitate tests for UWM
+func setupUWM(t *testing.T) error {
+	t.Helper()
+	ctx := context.Background()
+
+	cm := &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      clusterMonitorConfigMapName,
+			Namespace: f.Ns,
+		},
+		Data: map[string]string{
+			"config.yaml": `enableUserWorkload: true
+`,
+		},
+	}
+
+	uwmCM := &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      userWorkloadMonitorConfigMapName,
+			Namespace: f.UserWorkloadMonitoringNs,
+		},
+		Data: map[string]string{
+			"config.yaml": `prometheus:
+  volumeClaimTemplate:
+    spec:
+      resources:
+        requests:
+          storage: 2Gi
+`,
+		},
+	}
+
+	if err := f.OperatorClient.CreateOrUpdateConfigMap(ctx, cm); err != nil {
+		return err
+	}
+
+	if err := f.OperatorClient.CreateOrUpdateConfigMap(ctx, uwmCM); err != nil {
+		return err
+	}
+
+	// wait for required dependencies before executing tests
+	err := f.OperatorClient.WaitForStatefulsetRollout(ctx, &appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "prometheus-user-workload",
+			Namespace: f.UserWorkloadMonitoringNs,
+		},
+	})
+	if err != nil {
+		return err
+	}
+	if err := deployUserApplication(t); err != nil {
+		return err
+	}
+	return createPrometheusAlertmanagerInUserNamespace(t)
+}
+
+func deployUserApplication(t *testing.T) error {
```

Comment:
this could also be moved to the framework I think

Reply:
done